File: src/sc_neurocore/accel/go/services/gif_population.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SC-NeuroCore — Go service for gif_population

package services

import "math"

// GIFPopulationNeuronState holds the Mensi et al. GIF state and parameters.
type GIFPopulationNeuronState struct {
	V            float64
	Theta        float64
	Eta          float64
	TauM         float64
	TauEta       float64
	DeltaV       float64
	Lambda0      float64
	EtaIncrement float64
	VRest        float64
	VReset       float64
	Dt           float64
	Seed         uint64
	rng          uint64
}

// NewGIFPopulationNeuron creates a new GIFPopulationNeuron neuron with default parameters.
func NewGIFPopulationNeuron() *GIFPopulationNeuronState {
	return NewGIFPopulationNeuronWithSeed(42)
}

// NewGIFPopulationNeuronWithSeed creates a deterministic seeded GIF neuron.
func NewGIFPopulationNeuronWithSeed(seed uint64) *GIFPopulationNeuronState {
	if seed == 0 {
		seed = 1
	}
	return &GIFPopulationNeuronState{
		V:            -65.0,
		Theta:        -50.0,
		Eta:          0.0,
		TauM:         20.0,
		TauEta:       100.0,
		DeltaV:       2.0,
		Lambda0:      0.001,
		EtaIncrement: 5.0,
		VRest:        -65.0,
		VReset:       -65.0,
		Dt:           0.5,
		Seed:         seed,
		rng:          seed,
	}
}

func finiteGIFValues(values ...float64) bool {
	for _, value := range values {
		if math.IsNaN(value) || math.IsInf(value, 0) {
			return false
		}
	}
	return true
}

func (s *GIFPopulationNeuronState) validRuntime() bool {
	return finiteGIFValues(s.V, s.Theta, s.Eta, s.TauM, s.TauEta, s.DeltaV, s.Lambda0, s.EtaIncrement, s.VRest, s.VReset, s.Dt) &&
		s.TauM > 0.0 && s.TauEta > 0.0 && s.DeltaV > 0.0 && s.Lambda0 >= 0.0 && s.Dt > 0.0
}

func (s *GIFPopulationNeuronState) uniform() float64 {
	x := s.rng
	x ^= x >> 12
	x ^= x << 25
	x ^= x >> 27
	s.rng = x
	return float64((x*2685821657736338717)>>11) * (1.0 / 9007199254740992.0)
}

func (s *GIFPopulationNeuronState) advanceSubthreshold(iExt float64) (float64, float64, bool) {
	etaDecay := math.Exp(-s.Dt / s.TauEta)
	membraneDecay := math.Exp(-s.Dt / s.TauM)
	x0 := s.V - s.VRest - iExt
	etaNew := s.Eta * etaDecay
	var xNew float64
	if math.Abs(s.TauM-s.TauEta) <= 1e-12 {
		xNew = membraneDecay * (x0 - s.Eta*s.Dt/s.TauM)
	} else {
		coupling := s.TauEta / (s.TauEta - s.TauM)
		xNew = x0*membraneDecay - s.Eta*coupling*(etaDecay-membraneDecay)
	}
	vNew := s.VRest + iExt + xNew
	return vNew, etaNew, finiteGIFValues(vNew, etaNew)
}

func (s *GIFPopulationNeuronState) spikeProbability(voltage float64) float64 {
	if s.Lambda0 == 0.0 {
		return 0.0
	}
	exponent := math.Max(math.Min((voltage-s.Theta)/s.DeltaV, 20.0), -745.0)
	hazard := s.Lambda0 * math.Exp(exponent)
	pSpike := 1.0 - math.Exp(-hazard*s.Dt)
	return math.Max(math.Min(pSpike, 1.0), 0.0)
}

// Step advances the neuron by one timestep.
func (s *GIFPopulationNeuronState) Step(iExt float64) int {
	if math.IsNaN(iExt) || math.IsInf(iExt, 0) || !s.validRuntime() {
		return 0
	}
	vCandidate, etaCandidate, ok := s.advanceSubthreshold(iExt)
	if !ok {
		return 0
	}
	s.V = vCandidate
	s.Eta = etaCandidate
	if s.uniform() < s.spikeProbability(s.V) {
		s.V = s.VReset
		s.Eta += s.EtaIncrement
		return 1
	}
	return 0
}

// Reset restores the dynamic state and deterministic random stream.
func (s *GIFPopulationNeuronState) Reset() {
	s.V = s.VRest
	s.Eta = 0.0
	s.rng = s.Seed
}

// SimulateGIFPopulationNeuron runs the neuron for n steps.
func SimulateGIFPopulationNeuron(nSteps int, iExt float64) ([]float64, int) {
	s := NewGIFPopulationNeuron()
	trace := make([]float64, nSteps)
	spikes := 0
	for step := 0; step < nSteps; step++ {
		result := s.Step(iExt)
		trace[step] = s.V
		if result > 0 {
			spikes++
		}
	}
	return trace, spikes
}

File: src/sc_neurocore/accel/julia/neurons/gif_population.jl
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — Julia for gif_population

module GifPopulationAccel

export step!, simulate, reset!, GIFPopulationNeuronState

mutable struct GIFPopulationNeuronState
    v::Float64
    theta::Float64
    eta::Float64
    tau_m::Float64
    tau_eta::Float64
    delta_v::Float64
    lambda_0::Float64
    eta_increment::Float64
    v_rest::Float64
    v_reset::Float64
    dt::Float64
    seed::UInt64
    rng::UInt64
end

function GIFPopulationNeuronState(seed::UInt64=UInt64(42))
    normalized_seed = seed == 0 ? UInt64(1) : seed
    GIFPopulationNeuronState(-65.0, -50.0, 0.0, 20.0, 100.0, 2.0, 0.001, 5.0, -65.0, -65.0, 0.5, normalized_seed, normalized_seed)
end

function finite_values(values::Float64...)
    all(isfinite, values)
end

function valid_runtime(s::GIFPopulationNeuronState)
    finite_values(s.v, s.theta, s.eta, s.tau_m, s.tau_eta, s.delta_v, s.lambda_0, s.eta_increment, s.v_rest, s.v_reset, s.dt) &&
        s.tau_m > 0.0 && s.tau_eta > 0.0 && s.delta_v > 0.0 && s.lambda_0 >= 0.0 && s.dt > 0.0
end

function uniform!(s::GIFPopulationNeuronState)
    x = s.rng
    x ⊻= x >> 12
    x ⊻= x << 25
    x ⊻= x >> 27
    s.rng = x
    Float64((x * UInt64(2685821657736338717)) >> 11) * (1.0 / 9007199254740992.0)
end

function advance_subthreshold(s::GIFPopulationNeuronState, I_ext::Float64)
    eta_decay = exp(-s.dt / s.tau_eta)
    membrane_decay = exp(-s.dt / s.tau_m)
    x0 = s.v - s.v_rest - I_ext
    eta_new = s.eta * eta_decay
    if abs(s.tau_m - s.tau_eta) <= 1e-12
        x_new = membrane_decay * (x0 - s.eta * s.dt / s.tau_m)
    else
        coupling = s.tau_eta / (s.tau_eta - s.tau_m)
        x_new = x0 * membrane_decay - s.eta * coupling * (eta_decay - membrane_decay)
    end
    v_new = s.v_rest + I_ext + x_new
    return v_new, eta_new, finite_values(v_new, eta_new)
end

function spike_probability(s::GIFPopulationNeuronState, voltage::Float64)
    s.lambda_0 == 0.0 && return 0.0
    exponent = clamp((voltage - s.theta) / s.delta_v, -745.0, 20.0)
    hazard = s.lambda_0 * exp(exponent)
    clamp(1.0 - exp(-hazard * s.dt), 0.0, 1.0)
end

function step!(s::GIFPopulationNeuronState, I_ext::Float64=0.0; dt::Float64=s.dt)
    s.dt = dt
    if !isfinite(I_ext) || !valid_runtime(s)
        return 0
    end
    v_candidate, eta_candidate, ok = advance_subthreshold(s, I_ext)
    if !ok
        return 0
    end
    s.v = v_candidate
    s.eta = eta_candidate
    if uniform!(s) < spike_probability(s, s.v)
        s.v = s.v_reset
        s.eta += s.eta_increment
        return 1
    end
    return 0
end

function reset!(s::GIFPopulationNeuronState)
    s.v = s.v_rest
    s.eta = 0.0
    s.rng = s.seed
    return nothing
end

function simulate(n_steps::Int=1000; I_ext::Float64=10.0, dt::Float64=0.5)
    s = GIFPopulationNeuronState()
    trace = zeros(n_steps)
    spikes = 0
    for t in 1:n_steps
        result = step!(s, I_ext; dt=dt)
        trace[t] = s.v
        if result > 0
            spikes += 1
        end
    end
    return trace, spikes
end

end # module GifPopulationAccel
